src/vm/internal/otable.cc

```diff
@@ -1,271 +1,131 @@
-  while (*full == '/') {
-    full++;
-    (*size)--;
-  }
-  strcpy(name, full);
-  tmp = strchr(name, '#');
-  if (tmp) {
-    *size = tmp - name;
-  }
-  while (*size > 2 && name[*size - 1] == 'c' && name[*size - 2] == '.') {
-    *size -= 2;
-  }
-  name[*size] = 0;
-  name = extend_string(name, *size);
-  return name;
-}
-
-void init_otable() {
-  int x, y;
+#include <algorithm>
+#include <sstream>
 
-  /* ensure that otable_size is a power of 2 */
-  y = CONFIG_INT(__OBJECT_HASH_TABLE_SIZE__);
+ObjectTable::Singleton ObjectTable::instance_ = nullptr;
 
-  for (otable_size = 1; otable_size < y; otable_size *= 2) {
-    ;
-  }
-  otable_size_minus_one = otable_size - 1;
-  obj_table = reinterpret_cast<object_t **>(
-      DCALLOC(otable_size, sizeof(object_t *), TAG_OBJ_TBL, "init_otable"));
-  ch_table = reinterpret_cast<object_t **>(
-      DCALLOC(otable_size, sizeof(object_t *), TAG_OBJ_TBL, "init_ch_otable"));
+ObjectTable::ObjectTable()
+:objects_({}),children_({})
+{}
 
-  for (x = 0; x < otable_size; x++) {
-    obj_table[x] = 0;
-    ch_table[x] = 0;
-  }
+//static method to return a pointer to the singleton object table.
+ObjectTable::Singleton ObjectTable::get() {
+    if( instance_ == nullptr )
+        instance_ = Singleton( new ObjectTable() );
+    return instance_;
 }
 
-/*
- * Looks for obj in table, moves it to head.
- */
-
-static long obj_searches = 0, obj_probes = 0, objs_found = 0;
-
-/* A global.  *shhhh* don't tell. */
-static int h;
-static int ch;
-
-static object_t *find_obj_n(const char *s) {
-  object_t *curr, *prev;
-  h = ObjHash(s);
-  curr = obj_table[h];
-  prev = 0;
-
-  obj_searches++;
-
-  while (curr) {
-    obj_probes++;
-    if (!strcmp(curr->obname, s)) { /* found it */
-      if (prev) {                   /* not at head of list */
-        prev->next_hash = curr->next_hash;
-        curr->next_hash = obj_table[h];
-        obj_table[h] = curr;
-      }
-      objs_found++;
-      return (curr); /* pointer to object */
+//fix style and efficiency errors in this function
+//should pass in a K const& k. also why is V not prefixed by ObjectTable and
+//it still works? is the ObjectTable needed in the parameter list? 
+
+//attempt to insert an key(obname), object pointer pair into the object table if the key is not in the table.
+//also will insert the key, object pointer pair in the childrens list.
+//if the key is already in the object table it fails and returns false, otherwise true.
+bool ObjectTable::insert(Key const& key,Value value) {
+    if( objects_.find(key) == objects_.end() ) {
+        objects_.insert( std::make_pair(key,value) );
+        auto base = basename(key);
+        auto it = children_.find(base);
+        if( it == children_.end() ) {
+          children_.insert( std::pair<Key,Vector>(base, {value} ) );
+        }
+        else {
+            it->second.push_back(value);
+        }
+        return true;
+    } 
+    else {
+        return false;
```

Comment:
reverse the case here, always put the error case first , so you don't have to write a "else{}"